proc/cond.c

```diff
@@ -5,115 +5,129 @@
  *
  * Conditionals
  *
- * Copyright 2017 Phoenix Systems
- * Author: Pawel Pisarczyk
+ * Copyright 2017, 2023 Phoenix Systems
+ * Author: Pawel Pisarczyk, Aleksander Kaminski
  *
  * This file is part of Phoenix-RTOS.
  *
  * %LICENSE%
  */
 
 #include "hal/hal.h"
+#include "lib/assert.h"
 #include "include/errno.h"
 #include "threads.h"
 #include "cond.h"
 #include "mutex.h"
 #include "resource.h"
 
 
-int cond_put(cond_t *cond)
+cond_t *cond_get(int c)
 {
-	int rem;
-
-	if (!(rem = resource_put(&cond->resource)))
-		vm_kfree(cond);
-
-	return rem;
+	thread_t *t = proc_current();
+	resource_t *r = resource_get(t->process, c);
+	LIB_ASSERT((r == NULL) || (r->type == rtCond), "process: %s, pid: %d, tid: %d, handle: %d, resource type mismatch",
+		t->process->path, process_getPid(t->process), proc_getTid(t), c);
+	return ((r != NULL) && (r->type == rtCond)) ? r->payload.cond : NULL;
 }
 
 
-cond_t *cond_get(unsigned int c)
+void cond_put(cond_t *cond)
 {
-	return resourceof(cond_t, resource, resource_get(proc_current()->process, rtCond, c));
+	thread_t *t = proc_current();
+	int rem = resource_put(t->process, &cond->resource);
+	LIB_ASSERT(rem >= 0, "process: %s, pid: %d, tid: %d, refcnt below zero",
+		t->process->path, process_getPid(t->process), proc_getTid(t));
```

Comment:
What do you think about adding a check if the pointer `cond == NULL` at the begining of this function and removing check before calling this function.
Missing empty line after the declaration of local variables.
```suggestion
	thread_t *t;
	int rem;

	if (cond == NULL) {
		return;
	}
	t = proc_current();
	rem = resource_put(t->process, &cond->resource);
	LIB_ASSERT(rem >= 0, "process: %s, pid: %d, tid: %d, refcnt below zero",
		t->process->path, process_getPid(t->process), proc_getTid(t));
```

Reply:
But you can remove the check in the `userintr.c` file ( for times ) such as this:
```c
		if (cond != NULL) {
			cond_put(cond);
		}
```

Reply:
I'm aware, but it's a matter of - do we allow to `put` a resource you never had? IMHO it's a bad practice to do so and it makes a code using it somewhat less clear